In [1]:
import numpy as np
import pandas as pd
import datetime


In [24]:
def strip_strings(df,columns_to_strip):
    # strip strings
    df[columns_to_strip].str.strip()
    for elem in columns_to_strip:
      df[elem] = df[elem].str.strip()
     


def lower_strings(df):
    # strip strings
    columns_to_lower = ['curb_loc',
                        'status',
                        'health',
                        'spc_latin',
                        'spc_common',
                        'steward',
                        'guards',
                        'sidewalk',
                        'user_type',
                        'problems',
                        'root_stone',
                        'root_grate',
                        'root_other',
                        'trnk_wire',
                        'trnk_light',
                        'trnk_other',
                        'brnch_ligh',
                        'brnch_shoe',
                        'brnch_othe',
                        'address',
                        'zip_city',
                        'boroname']
    
    df = df[columns_to_lower].apply(lambda x: x.str.lower())
    return df

In [25]:

url='data_100000.csv'
df_original=pd.read_csv(url, sep=',')
#df_original.head()

#df.info()
# for security->df
df = df_original.copy()
#df.head()

#drop columns 
df.drop(columns=['the_geom','state','latitude','longitude'], inplace=True)
df.head()

#total_nan_values = df.isnull().sum().sum()
#print ("total_nan_values=",total_nan_values) #34945

#total_na_values = df.isna().sum().sum()
#print ("total_na_values=",total_na_values) #34945

#df=df.dropna()
df = df.dropna(how='all')

#convert date_time
df['created_at'] = pd.to_datetime(df['created_at'])
df.head()


created_at  tree_id  block_id  tree_dbh  stump_diam curb_loc status health  \
0 2015-08-27   180683    348711         3           0   OnCurb  Alive   Fair   
1 2015-09-03   200540    315986        21           0   OnCurb  Alive   Fair   
2 2015-09-05   204026    218365         3           0   OnCurb  Alive   Good   
3 2015-09-05   204337    217969        10           0   OnCurb  Alive   Good   
4 2015-08-30   189565    223043        21           0   OnCurb  Alive   Good   

                            spc_latin       spc_common  ... borocode  \
0                         Acer rubrum        red maple  ...        4   
1                   Quercus palustris          pin oak  ...        4   
2  Gleditsia triacanthos var. inermis      honeylocust  ...        3   
3  Gleditsia triacanthos var. inermis      honeylocust  ...        3   
4                     Tilia americana  American linden  ...        3   

   boroname cncldist st_assem st_senate   nta            nta_name  boro_ct  \
0    Queens       29       28        16  QN17        Forest Hills  4073900   
1    Queens       19       27        11  QN49          Whitestone  4097300   
2  Brooklyn       34       50        18  BK90   East Williamsburg  3044900   
3  Brooklyn       34       53        18  BK90   East Williamsburg  3044900   
4  Brooklyn       39       44        21  BK37  Park Slope-Gowanus  3016500   

           x_sp           y_sp  
0  1.027431e+06  202756.768749  
1  1.034456e+06  228644.837379  
2  1.001823e+06  200716.891267  
3  1.002420e+06  199244.253136  
4  9.909138e+05  182202.425999  

[5 rows x 38 columns]

In [27]:
#strip "info" in the columns
columns_to_strip = ['curb_loc',
                        'status',
                        'health',
                        'spc_latin',
                        'spc_common',
                        'steward',
                        'guards',
                        'sidewalk',
                        'user_type',
                        'problems',
                        'root_stone',
                        'root_grate',
                        'root_other',
                        'trnk_wire',
                        'trnk_light',
                        'trnk_other',
                        'brnch_ligh',
                        'brnch_shoe',
                        'brnch_othe',
                        'address',
                        'zip_city',
                        'boroname']

df['curb_loc'] = df['curb_loc'].str.strip()
df['status'] = df['status'].str.strip()
df['health'] = df['health'].str.strip()
df['spc_latin'] = df['spc_latin'].str.strip()
df['spc_common'] = df['spc_common'].str.strip()
df['steward'] = df['steward'].str.strip()
df['guards'] = df['guards'].str.strip()
df['sidewalk'] = df['sidewalk'].str.strip()
df['problems'] = df['problems'].str.strip()
df['root_stone'] = df['root_stone'].str.strip()
df['root_grate'] = df['root_grate'].str.strip()
df['root_other'] = df['root_other'].str.strip()
df['brnch_ligh'] = df['brnch_ligh'].str.strip()
df['brnch_shoe'] = df['brnch_shoe'].str.strip()
df['brnch_othe'] = df['brnch_othe'].str.strip()
df['address'] = df['address'].str.strip()
df['zip_city'] = df['zip_city'].str.strip()
df['boroname'] = df['boroname'].str.strip()

#lower_strings(df): (e.g. `Truck` and `truck` refer to the same thing)
df['curb_loc'] = df['curb_loc'].str.lower()
df['status'] = df['status'].str.lower()
df['health'] = df['health'].str.lower()
df['spc_latin'] = df['spc_latin'].str.lower()
df['spc_common'] = df['spc_common'].str.lower()
df['steward'] = df['steward'].str.lower()
df['guards'] = df['guards'].str.lower()
df['sidewalk'] = df['sidewalk'].str.lower()
df['problems'] = df['problems'].str.lower()
df['root_stone'] = df['root_stone'].str.lower()
df['root_grate'] = df['root_grate'].str.lower()
df['root_other'] = df['root_other'].str.lower()
df['brnch_ligh'] = df['brnch_ligh'].str.lower()
df['brnch_shoe'] = df['brnch_shoe'].str.lower()
df['brnch_othe'] = df['brnch_othe'].str.lower()
df['address'] = df['address'].str.lower()
df['zip_city'] = df['zip_city'].str.lower()
df['boroname'] = df['boroname'].str.lower()

#all column change from bool to int
columns_to_int = ['root_stone',
                      'root_grate',
                      'root_other',
                      'trnk_wire',
                      'trnk_light',
                      'trnk_other',
                      'brnch_ligh',
                      'brnch_shoe',
                      'brnch_othe']

df['root_stone'] = df['root_stone']*1
df['root_other'] = df['root_other']*1
df['trnk_wire'] = df['trnk_wire']*1
df['trnk_light'] = df['trnk_light']*1
df['trnk_other'] = df['trnk_other']*1
df['brnch_ligh'] = df['brnch_ligh']*1
df['brnch_ligh'] = df['brnch_ligh']*1
df['brnch_shoe'] = df['brnch_shoe']*1
df['othetone'] = df['brnch_othe']*1

df.head()


created_at  tree_id  block_id  tree_dbh  stump_diam curb_loc status health  \
0 2015-08-27   180683    348711         3           0   oncurb  alive   fair   
1 2015-09-03   200540    315986        21           0   oncurb  alive   fair   
2 2015-09-05   204026    218365         3           0   oncurb  alive   good   
3 2015-09-05   204337    217969        10           0   oncurb  alive   good   
4 2015-08-30   189565    223043        21           0   oncurb  alive   good   

                            spc_latin       spc_common  ...  boroname  \
0                         acer rubrum        red maple  ...    queens   
1                   quercus palustris          pin oak  ...    queens   
2  gleditsia triacanthos var. inermis      honeylocust  ...  brooklyn   
3  gleditsia triacanthos var. inermis      honeylocust  ...  brooklyn   
4                     tilia americana  american linden  ...  brooklyn   

  cncldist st_assem st_senate   nta            nta_name  boro_ct  \
0       29       28        16  QN17        Forest Hills  4073900   
1       19       27        11  QN49          Whitestone  4097300   
2       34       50        18  BK90   East Williamsburg  3044900   
3       34       53        18  BK90   East Williamsburg  3044900   
4       39       44        21  BK37  Park Slope-Gowanus  3016500   

           x_sp           y_sp othetone  
0  1.027431e+06  202756.768749       no  
1  1.034456e+06  228644.837379       no  
2  1.001823e+06  200716.891267       no  
3  1.002420e+06  199244.253136       no  
4  9.909138e+05  182202.425999       no  

[5 rows x 39 columns]

In [28]:
print('NaN occurrences in Columns:')
print(df.isna().sum())


NaN occurrences in Columns:
created_at       0
tree_id          0
block_id         0
tree_dbh         0
stump_diam       0
curb_loc         0
status           0
health        4993
spc_latin     4992
spc_common    4992
steward       4992
guards        4992
sidewalk      4992
user_type        0
problems      4992
root_stone       0
root_grate       0
root_other       0
trnk_wire        0
trnk_light       0
trnk_other       0
brnch_ligh       0
brnch_shoe       0
brnch_othe       0
address          0
zipcode          0
zip_city         0
cb_num           0
borocode         0
boroname         0
cncldist         0
st_assem         0
st_senate        0
nta              0
nta_name         0
boro_ct          0
x_sp             0
y_sp             0
othetone         0
dtype: int64


In [29]:
print('NaN occurrences in Rows:')
df_row_nan=df[df.isnull().sum(axis = 1)>1]
print(df_row_nan)

NaN occurrences in Rows:
      created_at  tree_id  block_id  tree_dbh  stump_diam curb_loc status  \
630   2015-08-31   192569    302968         2           0   oncurb   dead   
631   2015-08-27   179766    230466         0          30   oncurb  stump   
634   2015-08-27   180619    348834         0          10   oncurb  stump   
635   2015-08-29   184295    505610         0           0   oncurb   dead   
636   2015-09-08   208828    107238         0          15   oncurb  stump   
...          ...      ...       ...       ...         ...      ...    ...   
99291 2015-10-08   296790    230356         0          34   oncurb  stump   
99293 2015-10-10   303810    111012         0          15   oncurb  stump   
99294 2015-10-08   297660    230365         0          12   oncurb  stump   
99295 2015-10-09   303165    226311         0           7   oncurb  stump   
99296 2015-10-10   306550    104059         4           0   oncurb   dead   

      health spc_latin spc_common  ...   boroname 

#### After looking the Nan values in 'columns' and 'rows', we are sure hier that all the Nan values are located in the 4992 rows.
##### I think perhaps these are "new_categories" because all the columns that we don't have informations, they are in the 
##### categorie "name" (health, spc_latin, spc_common,steward, guards, sidewalk, problems,..). I propose to create the "new_categories" and ##### don't delete them.


In [30]:
#write to new csv final:
df.to_csv("data_100000_clean.csv")